File: lintcode/12_min-stack/min-stack.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/min-stack
@Language: Python
@Datetime: 16-04-29 00:31
'''

class MinStack(object):

    def __init__(self):
        self.stk1 = []
        self.stk2 = []

    def push(self, number):
        self.stk1.append(number)
        self.stk2.append(number if not self.stk2 else min(self.stk2[-1], number))

    def pop(self):
        self.stk2.pop()
        return self.stk1.pop()

    def min(self):
        return self.stk2[-1] if self.stk2 else None

File: lintcode/12_min-stack/min-stack.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/min-stack
@Language: C++
@Datetime: 16-07-21 23:59
*/

class MinStack {
    stack<int> stk;
    stack<int> stk_min;
public:
    MinStack() {}

    void push(int number) {
        stk.push(number);
        if (stk_min.empty() || number <= stk_min.top())
            stk_min.push(number);
        else
            stk_min.push(stk_min.top());
    }

    int pop() {
        int top = stk.top();
        stk.pop();
        stk_min.pop();
        return top;
    }

    int min() {
        return stk_min.top();
    }
};
